src/voice/connection.rs

```diff
@@ -374,95 +406,111 @@ impl Connection {
+                        .or_default()
+                        .push(i);
                 }
+            
+                i += 1;
 
-                // Assume this for now, at least.
-                // We'll be fusing streams, so we can either keep
-                // as stereo or downmix to mono.
-                let is_stereo = true;
-                let source_stereo = stream.is_stereo();
-
-                if is_stereo != self.encoder_stereo {
-                    let channels = if is_stereo {
-                        Channels::Stereo
-                    } else {
-                        Channels::Mono
-                    };
-                    self.encoder = OpusEncoder::new(SAMPLE_RATE, channels, CodingMode::Audio)?;
-                    self.encoder_stereo = is_stereo;
-                }
-
-                let temp_len = match stream.get_type() {
-                    AudioType::Opus => if stream.decode_and_add_opus_frame(&mut mix_buffer, vol).is_some() {
-                            opus_frame.len()
-                        } else {
-                            0
-                        },
-                    AudioType::Pcm => {
-                        let buffer_len = if source_stereo { 960 * 2 } else { 960 };
-
-                        match stream.read_pcm_frame(&mut buffer[..buffer_len]) {
-                            Some(len) => len,
-                            None => 0,
-                        }
-                    },
-                    AudioType::__Nonexhaustive => unreachable!(),
-                };
+                continue;
+            }
 
-                // May need to force interleave/copy.
-                combine_audio(*buffer, &mut mix_buffer, source_stereo, vol);
+            // Assume this for now, at least.
+            // We'll be fusing streams, so we can either keep
+            // as stereo or downmix to mono.
+            let is_stereo = true;
 
-                len = len.max(temp_len);
-                i += if temp_len > 0 {
-                    1
+            if is_stereo != self.encoder_stereo {
+                let channels = if is_stereo {
+                    Channels::Stereo
                 } else {
-                    sources.remove(i);
-                    finished = true;
-
-                    0
+                    Channels::Mono
                 };
+                self.encoder = OpusEncoder::new(SAMPLE_RATE, channels, CodingMode::Audio)?;
+                self.encoder_stereo = is_stereo;
             }
 
-            aud.finished = finished;
+            let now = Instant::now();
+            let temp_len = stream.mix(&mut mix_buffer, vol);
+            let later = Instant::now();
 
-            if !finished {
+            *time_in_call += later - now;
+            *entry_points += 1;
+
+            if *entry_points % 1000 == 0 {
+                println!("Average cost {:?}ns", time_in_call.as_nanos()/(*entry_points as u128));
```

Comment:
This should likely be a `debug!()`, `trace!()` or possibly `info!()`

Reply:
This (and `entry_points`) will be removed once the patch is finalised.